### PR TITLE
features/MultiLine Form Field

### PR DESCRIPTION
**Pull request by ibelar**
## MultiLine field

New form input in order to allow add/edit multiple record within a form.

## Usage

```
$f = $app->add('Form');
$f->setModel($invoice, false);

$sub_layout = $f->layout->addSubLayout('Generic');
$sub_layout->setModel($invoice, ['name', 'client_id', 'address']);

$sub_layout = $f->layout->addSubLayout('Generic');

//Add multi line input to the layout.
$ml = $sub_layout->addField('ml', ['MultiLine']);
$ml->setModel($invoice, ['item','cat','qty','price', 'sub', 'tx', 'grd'], 'Items', 'invoice_id');

// add onchange handler and update grand total.
$ml->onChange(function($rows) use ($f) {
    $total = 0;
    foreach ($rows as $row => $cols) {
        foreach ($cols as $col) {
            $fieldName = key($col);
            if ($fieldName === 'grd') {
                $total = $total + substr($col[$fieldName], 4);
            }
        }
    }
 return $f->js(true, null, 'input[name="g_total"]')->val(number_format($total, 2));

}, ['qty', 'price']);
```

Screen Shot 2019-05-03 at 11 06 02 AM

## Demo file
multiline.php

## Feature

- Adding row will use Model/Field default value if set.
- Multiple row deletion;
- Calculate Expression or Callback value from Field definition;
- onChange callback for entire row record or on specify field.
- Validate according to Field definition in model.
- Validation is apply on every row when form is saved.

Screen Shot 2019-05-03 at 11 06 31 AM

### CODE REVIEW

**Comment by romaninsh** [comment]
I'd like $rows to be a model linked with array persistence. Also - user should be able to change that with effect on the UI. 

@ibelar are you sending ALL the rows back to the frontend or just changed ones ?

**Comment by romaninsh** [comment]
I'd like to play with this implementation a little bit before merging.

File: src/FormField/MultiLine.php [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+                }
+                $value = $col[$fieldName];
+
+                try {
+                    $field = $m->getElement($fieldName);
+                    // save field value only if field was editable in form at all
+                    if (!$field->read_only) {
+                        $m[$fieldName] = $this->app->ui_persistence->typecastLoadField($field, $value);
+                    }
+                } catch (\atk4\core\Exception $e) {
```
Comment (romaninsh) [inline src/FormField/MultiLine.php]: we should only catch and display validation errors (they have a dedicated class).

File: src/FormField/MultiLine.php [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+    /**
+     * Render callback.
+     *
+     * @throws ValidationException
+     * @throws \atk4\core\Exception
+     * @throws \atk4\data\Exception
+     */
+    private function renderCallback()
+    {
+        $action = isset($_POST['action']) ? $_POST['action'] : null;
```
Comment (romaninsh) [inline src/FormField/MultiLine.php]: could this be using callback's trigger value?

File: src/FormField/MultiLine.php [hunk 3]
```diff
@@ -0,0 +1,764 @@
...
+     *
+     * @param $exprField
+     * @param $fieldName
+     *
+     * @return int|mixed|string
+     */
+    private function getValueForExpression($exprField, $fieldName, $model)
+    {
+        switch ($exprField->type) {
+            case 'money':
```
Comment (romaninsh) [inline src/FormField/MultiLine.php]: ok, FOR NOW, but would have to refactor.

File: src/FormField/MultiLine.php [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+     * @throws \atk4\core\Exception
+     *
+     * @return mixed
+     */
+    private function getDummyExpression($exprField, $model)
+    {
+        $expr = $exprField->expr;
+        $matches = [];
+
+        preg_match_all('/\[[a-z0-9_]*\]|{[a-z0-9_]*}/i', $expr, $matches);
```
Comment (romaninsh) [inline src/FormField/MultiLine.php]: rather should rely on Model's callbacks.. OK for now. Or rather Persistence should do this.

File: src/FormField/MultiLine.php (hunk 1 above)
Comment (ibelar) [reply src/FormField/MultiLine.php]: Was not sure how to check for post load validation error. Thought that typecastLoadField was supposed to do so. Let me know how this should work and I can change it.

File: src/FormField/MultiLine.php (hunk 2 above)
Comment (ibelar) [reply src/FormField/MultiLine.php]: The callback url is render only once for the component but the same callback url is used for two different action.

**Comment by romaninsh** [comment]
can't merge, CI fails.

**Comment by ibelar** [comment]
@romaninsh - ci should passed now. Rename onChange method to onLineChange...

**Comment by romaninsh** [comment]
Thanks!

**Pull request merged**